### ISSUE TITLE
Phase 45.1: KnowledgeDistiller — Teacher-Student Knowledge Transfer

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#889]
### Overview
Implements teacher-student knowledge distillation framework enabling large model knowledge transfer to compact student networks while preserving accuracy. Supports multiple distillation strategies including soft target transfer, feature matching, attention transfer, and contrastive representation distillation.

#### 1. Temperature-Scaled Soft Target Distillation (Hinton et al. 2015)
- KL divergence loss between teacher/student softmax outputs
- Configurable temperature parameter τ (typically 2-20)
- Combined loss: `L = α·L_hard + (1-α)·T²·L_KL(soft_teacher, soft_student)`
- Adaptive temperature scheduling during training

#### 2. Feature Matching (FitNets, Romero et al. 2015)
- Intermediate layer hint-based training
- Learnable projection layers for dimension matching
- L2 and cosine similarity feature matching losses
- Multi-layer feature alignment with configurable matching points

#### 3. Attention Transfer (Zagoruyko & Komodakis 2017)
- Attention map extraction from activation tensors
- Channel-wise and spatial attention maps
- Activation-based attention: `A = Σ|F_i|^p` across channel dimension
- Gradient-weighted attention transfer

#### 4. Contrastive Representation Distillation (Tian et al. 2020)
- Cross-modal contrastive learning between teacher/student representations
- Memory buffer for negative sampling
- InfoNCE loss for structural knowledge transfer
- Momentum-updated representation bank

#### 5. Progressive Distillation
- Multi-step teacher chain: Teacher → TA₁ → TA₂ → Student
- Curriculum-based distillation scheduling
- Self-distillation (Born-Again Networks)
- Online distillation with mutual learning

### API Design

```python
class KnowledgeDistiller:
    def __init__(self, config: DistillationConfig):
        """Initialize distiller with strategy configuration."""
    
    def distill(
        self,
        teacher: nn.Module,
        student: nn.Module, 
        train_loader: DataLoader,
        strategy: str = "soft_target",  # soft_target|feature_match|attention|crd
        temperature: float = 4.0,
        alpha: float = 0.7,
        epochs: int = 100
    ) -> CompressedModel:
        """Execute knowledge distillation training."""
    
    def extract_knowledge(
        self,
        teacher: nn.Module,
        data: DataLoader
    ) -> KnowledgeRepresentation:
        """Extract transferable knowledge from teacher model."""
    
    def progressive_distill(
        self,
        teacher: nn.Module,
        student_chain: List[nn.Module],
        data: DataLoader
    ) -> CompressedModel:
        """Multi-step progressive distillation through teacher assistants."""
```

### Acceptance Criteria
- [ ] Soft target distillation with configurable temperature
- [ ] FitNet-style feature matching with projection layers
- [ ] Attention transfer with spatial and channel attention maps
- [ ] CRD with contrastive loss and memory buffer
- [ ] Progressive distillation pipeline
- [ ] Student accuracy within 2% of teacher on benchmark tasks
- [ ] Unit tests with >90% coverage
- [ ] Type annotations and comprehensive docstrings

### References
- Hinton et al. (2015) "Distilling the Knowledge in a Neural Network"
- Romero et al. (2015) "FitNets: Hints for Thin Deep Nets"
- Zagoruyko & Komodakis (2017) "Paying More Attention to Attention"
- Tian et al. (2020) "Contrastive Representation Distillation"
- Furlanello et al. (2018) "Born Again Neural Networks"

> Part of [Phase 45 Planning](#888)
